### PR TITLE
GH-1182 Port NTP Action Plugin to Rust

### LINKED ISSUE #1182: Port NTP Action Plugin to Rust
The NTP action plugin is currently written in the python backend. The purpose of this plugin is to synchronize time between the cluster nodes. Porting the NTP plugin will include the following:

1. Move logic from manage_ntp.py to a rust iml-agent action plugin
1. Create a daemon plugin that will check the ntp status when the manager sends its time. The agent will respond back with the agent time. 
1. Create a host plugin that will raise an alert if the ntp sync falls below a certain threshold. 
1. Create an action to resync the time
1. Add ability to configure ntp from iml cli

### PR DESCRIPTION
**Pull request by johnsonw**
Fixes #1182.

Port NTP Action Plugin to rust

Signed-off-by: Will Johnson <wjohnson@whamcloud.com>

### CODE REVIEW

**Comment by johnsonw** [comment]
Video of IML configuring NTP with the rust action plugin:

configure_ntp_action_plugin

File: iml-agent/Cargo.toml [hunk 1]
```diff
@@ -47,6 +47,7 @@ iml-fs = { path = "../iml-fs", version = "0.1.0" }
 [dev-dependencies]
 mockito = "0.20.0"
 pretty_assertions = "0.6.1"
+insta = "0.11.0"
```
Comment (jgrund) [inline iml-agent/Cargo.toml]: Use `0.11` here so patch changes are pulled in.

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+}
+
+/// Writes the new config data to the config file
+pub fn update_and_write_new_config(
+    server: Option<String>,
+) -> impl Future<Item = (), Error = ImlAgentError> {
+    let s = get_ntp_config_stream();
+    configure_ntp(server, s)
+        .and_then(|updated_config| tokio::fs::write(NTP_CONFIG_FILE, updated_config).from_err())
+        .then(|_| Ok(()))
```
Comment (jgrund) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]: Use `.map(drop)` here instead which won't swallow the error half of the future.

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        .fold(
+            ("".to_string(), false),
+            move |(mut acc, prefix_found), line| {
+                let (x, prefix_found) = if let Some(server_name) = &server {
+                    transform_config(server_name.into(), line, prefix_found)
+                } else {
+                    (line, prefix_found)
+                };
+
+                acc.push_str(format!("{}\n", x).as_ref());
```
Comment (jgrund) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]:
```suggestion
                acc.push_str(&format!("{}\n", x));
```

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+                    format!(
+                        "{}\n{}",
+                        format!("server  127.127.1.0 {}", MARKER),
+                        format!("fudge   127.127.1.0 stratum 10 {} {}", MARKER, line),
+                    ),
+                    true,
+                )
+            } else {
+                (
+                    [PREFIX, server.as_ref(), "iburst", MARKER, line.as_ref()].join(" "),
```
Comment (jgrund) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]:
```suggestion
                    [PREFIX, &server, "iburst", MARKER, &line].join(" "),
```

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+                    true,
+                )
+            }
+        }
+        (["server"], true) => ([MARKER, line.as_ref()].join(" ").to_string(), true),
```
Comment (jgrund) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]:
```suggestion
        (["server"], true) => ([MARKER, &line].join(" ").to_string(), true),
```

File: iml-agent/Cargo.toml (hunk 1 above)
Comment (johnsonw) [reply iml-agent/Cargo.toml]: 👍

File: iml-agent/src/action_plugins/ntp/action_configure.rs (hunk 2 above)
Comment (johnsonw) [reply iml-agent/src/action_plugins/ntp/action_configure.rs]: This function changed to:
```
pub fn get_ntp_config_stream() -> impl Stream<Item = String, Error = ImlAgentError> {
    iml_fs::stream_file_lines(Path::new(NTP_CONFIG_FILE))
        .map_err(|_| panic!("Failed to get a stream to the ntp config file."))
}
```

Comment (johnsonw) [reply iml-agent/src/action_plugins/ntp/action_configure.rs]: Should I still use a .map(drop)?

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+use futures::{future, Future, Stream};
+use std::path::Path;
+
+static NTP_CONFIG_FILE: &'static str = "/etc/ntp.conf";
+static MARKER: &'static str = "# IML_EDIT";
+static REMOVE_MARKER: &'static str = "#REMOVE_MARKER#";
+static PREFIX: &'static str = "server";
+
+/// Gets a stream to the ntp config
+pub fn get_ntp_config_stream() -> impl Stream<Item = String, Error = ImlAgentError> {
```
Comment (utopiabound) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]: Does this need to be public?

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+restrict 127.0.0.1 
+restrict ::1
+
+# Hosts on local network are less restricted.
+#restrict 192.168.1.0 mask 255.255.255.0 nomodify notrap
+
+# Use public servers from the pool.ntp.org project.
+# Please consider joining the pool (http://www.pool.ntp.org/join.html).
+server  127.127.1.0 # IML_EDIT
+fudge   127.127.1.0 stratum 10 # IML_EDIT server 0.centos.pool.ntp.org iburst
```
Comment (utopiabound) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]: this doesn't look right, should it be:
```
fudge   127.127.1.0 stratum 10 # IML_EDIT
# IML_EDIT server 0.centos.pool.ntp.org iburst
```

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+# the administrative functions.
+restrict 127.0.0.1 
+restrict ::1
+
+# Hosts on local network are less restricted.
+#restrict 192.168.1.0 mask 255.255.255.0 nomodify notrap
+
+# Use public servers from the pool.ntp.org project.
+# Please consider joining the pool (http://www.pool.ntp.org/join.html).
+server adm.local iburst # IML_EDIT server 0.centos.pool.ntp.org iburst
```
Comment (utopiabound) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]: same as above

File: iml-agent/src/action_plugins/ntp/action_configure.rs (hunk 2 above)
Comment (jgrund) [reply iml-agent/src/action_plugins/ntp/action_configure.rs]: Yes, using `then` will swallow any errors.

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+fn filter_out_remove_markers(line: &String) -> bool {
+    line.find(REMOVE_MARKER).is_none()
+}
+
+#[cfg(test)]
+mod test {
+    extern crate futures;
+    extern crate tokio;
```
Comment (jgrund) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]: Shouldn't need these

File: iml-agent/src/action_plugins/ntp/action_configure.rs (hunk 6 above)
Comment (johnsonw) [reply iml-agent/src/action_plugins/ntp/action_configure.rs]: removed pub

File: chroma_core/models/ntp.py [hunk 10]
```diff
@@ -45,7 +45,37 @@ def run(self, kwargs):
         else:
             ntp_server = socket.getfqdn()
 
-        self.invoke_agent_expect_result(kwargs["ntp_configuration"].host, "configure_ntp", {"ntp_server": ntp_server})
+        result = self.invoke_rust_agent_expect_result(kwargs["fqdn"], "configure_ntp", ntp_server)
```
Comment (utopiabound) [inline chroma_core/models/ntp.py]: Feels like this should just be:
```python
return self.invoke_rust_agent_expect_result(kwargs["fqdn"], "configure_ntp", ntp_server)
```

File: chroma_core/models/ntp.py [hunk 11]
```diff
@@ -89,7 +127,9 @@ class UnconfigureNTPStep(Step):
     idempotent = True
 
     def run(self, kwargs):
-        self.invoke_agent_expect_result(kwargs["ntp_configuration"].host, "unconfigure_ntp")
+        result = self.invoke_rust_agent_expect_result(kwargs["ntp_configuration"].host.fqdn, "configure_ntp", None)
+
+        return result
```
Comment (utopiabound) [inline chroma_core/models/ntp.py]: same comment as above

File: chroma_core/models/ntp.py (hunk 10 above)
Comment (jgrund) [reply chroma_core/models/ntp.py]: Updated.

File: chroma_core/models/ntp.py (hunk 11 above)
Comment (jgrund) [reply chroma_core/models/ntp.py]: Updated.

File: iml-agent/src/action_plugins/ntp/action_configure.rs [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+        .try_fold(
+            ("".to_string(), false),
+            move |(mut acc, prefix_found), line| {
+                let (x, prefix_found) = if let Some(server_name) = &server {
+                    transform_config(server_name.into(), line, prefix_found)
+                } else {
+                    (line, prefix_found)
+                };
+
+                acc.push_str(&format!("{}\n", x));
```
Comment (mkpankov) [inline iml-agent/src/action_plugins/ntp/action_configure.rs]: I think `push_str(x)` followed by `push_char('\n')` should be cheaper. Avoid allocations.

Comment (jgrund) [reply iml-agent/src/action_plugins/ntp/action_configure.rs]: Updated.

**Pull request merged**